### PR TITLE
[Merged by Bors] - chore: move to v4.8.0-rc1

### PR DESCRIPTION
**Pull request by kim-em**
This moves Mathlib to v4.8.0-rc1

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/AffineScheme.lean [hunk 1]
```diff
@@ -245,6 +245,8 @@ theorem isBasis_basicOpen (X : Scheme) [IsAffine X] :
 
 namespace IsAffineOpen
 
+-- Adaptation note: 2024-04-23
+-- TODO check if this `.{u}` restricts anything
```
Comment (semorrison) [inline Mathlib/AlgebraicGeometry/AffineScheme.lean:249]: Remove this?

File: Mathlib/Algebra/Order/GroupWithZero/Canonical.lean [hunk 2]
```diff
@@ -273,10 +273,15 @@ theorem OrderIso.mulRight₀'_symm {a : α} (ha : a ≠ 0) :
   rfl
 #align order_iso.mul_right₀'_symm OrderIso.mulRight₀'_symm
 
+set_option backward.isDefEq.lazyProjDelta false in -- See https://github.com/leanprover-community/mathlib4/issues/12535
```
Comment (semorrison) [inline Mathlib/Algebra/Order/GroupWithZero/Canonical.lean:276]: Duplicative with the in-line set_option?

File: Mathlib/Computability/AkraBazzi/AkraBazzi.lean [hunk 3]
```diff
@@ -1156,6 +1156,11 @@ lemma base_nonempty {n : ℕ} (hn : 0 < n) : (Finset.Ico (⌊b (min_bi b) / 2 *
                                  _ ≤ 1 * n        := by gcongr; norm_num
                                  _ = n             := by simp
 
+-- Adaptation note: 2024-04-23
+-- None of the backwards compatibility flags help here,
+-- but I think this timeout is the proof's fault for being too long, not Lean's fault.
+-- Refactor requested at https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Computability.2FAkraBazzi
+set_option maxHeartbeats 400000 in
```
Comment (semorrison) [inline Mathlib/Computability/AkraBazzi/AkraBazzi.lean:1163]: Now that #12541 has landed, hopefully we can remove this.

File: Mathlib/Data/Bool/AllAny.lean [hunk 4]
```diff
@@ -3,8 +3,10 @@ Copyright (c) 2017 Mario Carneiro. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Mario Carneiro
 -/
+import Mathlib.Data.Bool.Basic
 import Mathlib.Data.List.Basic
 import Mathlib.Init.Data.Bool.Lemmas
+import Mathlib.Data.Bool.Basic
```
Comment (semorrison) [inline Mathlib/Data/Bool/AllAny.lean:9]:
```suggestion
```

File: Mathlib/Data/Finset/Basic.lean [hunk 5]
```diff
@@ -2618,7 +2620,6 @@ theorem filter_comm (s : Finset α) : (s.filter p).filter q = (s.filter q).filte
 #align finset.filter_comm Finset.filter_comm
 
 -- We can simplify an application of filter where the decidability is inferred in "the wrong way"
-@[simp]
```
Comment (semorrison) [inline Mathlib/Data/Finset/Basic.lean:2621]: I've forgotten why/when this was removed.

Comment (semorrison) [reply Mathlib/Data/Finset/Basic.lean:2621]: It does seem excessively opinionated as a simp lemma, though. I'm happy it's removed. If someone is excited to try backporting to master?

File: Mathlib/Data/Fintype/BigOperators.lean [hunk 6]
```diff
@@ -176,12 +176,15 @@ end Fintype
...
-@[simp]
 theorem Fintype.card_fun [DecidableEq α] [Fintype α] [Fintype β] :
     Fintype.card (α → β) = Fintype.card β ^ Fintype.card α := by
   rw [Fintype.card_pi, Finset.prod_const]; rfl
 #align fintype.card_fun Fintype.card_fun
 
+@[simp]
+theorem Fintype.card_exp_univ_card_eq [Fintype α] [Fintype β] :
+    card β ^ univ.card (α := α) = card β ^ card α := rfl
+
```
Comment (semorrison) [inline Mathlib/Data/Fintype/BigOperators.lean:187]: Again, backport these? Or are the merge mistakes?

File: Mathlib/Data/List/Basic.lean [hunk 7]
```diff
@@ -1234,6 +1233,9 @@ theorem get?_length (l : List α) : l.get? l.length = none := get?_len_le le_rfl
 theorem nthLe_mem (l : List α) (n h) : nthLe l n h ∈ l := get_mem ..
 #align list.nth_le_mem List.nthLe_mem
 
+theorem nthLe_congr {l : List α} {n p : ℕ} {h : n < length l} (hnp : n = p) :
+    nthLe l n h = nthLe l p (hnp ▸ h) := by simp [hnp]
+
```
Comment (semorrison) [inline Mathlib/Data/List/Basic.lean:1238]: What's the deal here?

File: Mathlib/Data/Nat/Bitwise.lean [hunk 8]
```diff
@@ -3,11 +3,13 @@ Copyright (c) 2020 Markus Himmel. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Markus Himmel, Alex Keizer
 -/
+import Mathlib.Data.Bool.Basic
 import Mathlib.Data.List.GetD
 import Mathlib.Data.Nat.Bits
 import Mathlib.Algebra.Ring.Nat
 import Mathlib.Order.Basic
 import Mathlib.Tactic.Common
+import Mathlib.Data.Bool.Basic
```
Comment (semorrison) [inline Mathlib/Data/Nat/Bitwise.lean:12]:
```suggestion
```

File: Mathlib/Data/Fintype/BigOperators.lean (hunk 6 above)
Comment (semorrison) [reply Mathlib/Data/Fintype/BigOperators.lean:187]: I'm testing reverting these now.

File: Mathlib/Data/List/Basic.lean (hunk 7 above)
Comment (semorrison) [reply Mathlib/Data/List/Basic.lean:1238]: This is the sole appearance of this theorem, so I suspect its presence here is a merge mistake?

File: Mathlib/Data/Nat/Cast/Defs.lean [hunk 9]
```diff
@@ -156,8 +156,10 @@ theorem cast_one [AddMonoidWithOne R] : ((1 : ℕ) : R) = 1 := by
 #align nat.cast_one Nat.cast_oneₓ
 
 @[simp, norm_cast]
-theorem cast_add [AddMonoidWithOne R] (m n : ℕ) : ((m + n : ℕ) : R) = m + n := by
-  induction n <;> simp [add_succ, add_assoc, Nat.add_zero, Nat.cast_one, Nat.cast_zero, *]
+theorem cast_add {R : Type*} [AddMonoidWithOne R] (m n : ℕ) : ((m + n : ℕ) : R) = m + n := by
```
Comment (semorrison) [inline Mathlib/Data/Nat/Cast/Defs.lean:159]:
```suggestion
theorem cast_add [AddMonoidWithOne R] (m n : ℕ) : ((m + n : ℕ) : R) = m + n := by
```

File: Mathlib/Geometry/Manifold/VectorBundle/FiberwiseLinear.lean [hunk 10]
```diff
@@ -223,6 +223,13 @@ theorem SmoothFiberwiseLinear.locality_aux₂ (e : PartialHomeomorph (B × F) (B
 
 variable (F B IB)
 
+-- Adaptation note: 2024-04-23
+-- This `maxHeartbeats` was not previously required; increased from about 90k to about 320k.
+-- The backwards compatibility flags do not help here:
+-- set_option backward.isDefEq.lazyProjDelta false in
+-- set_option backward.isDefEq.lazyWhnfCore false in
+-- set_option backward.synthInstance.canonInstances false in
+set_option maxHeartbeats 400000 in
```
Comment (semorrison) [inline Mathlib/Geometry/Manifold/VectorBundle/FiberwiseLinear.lean:232]:
```suggestion
-- See https://leanprover.zulipchat.com/#narrow/stream/428973-nightly-testing/topic/Mathlib.20status.20updates/near/436257970
set_option maxHeartbeats 400000 in
```

File: Mathlib/Data/List/Basic.lean (hunk 7 above)
Comment (semorrison) [reply Mathlib/Data/List/Basic.lean:1238]: I'll remove this assuming things build locally.

File: Mathlib/Data/Nat/Dist.lean [hunk 11]
```diff
@@ -21,9 +21,8 @@ def dist (n m : ℕ) :=
   n - m + (m - n)
 #align nat.dist Nat.dist
 
-theorem dist.def (n m : ℕ) : dist n m = n - m + (m - n) :=
-  rfl
-#align nat.dist.def Nat.dist.def
+-- Should be aligned to `Nat.dist.eq_def`, but that is generated on demand and isn't present yet.
+#noalign nat.dist.eq_def
```
Comment (semorrison) [inline Mathlib/Data/Nat/Dist.lean:25]:
```suggestion
#noalign nat.dist.def
```

File: Mathlib/Tactic/CC/Datatypes.lean [hunk 12]
```diff
@@ -4,11 +4,12 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Leonardo de Moura, Miyahara Kō
 -/
 import Lean.Meta.CongrTheorems
-import Std.Tactic.Relation.Rfl
+import Lean.Meta.Tactic.Rfl
 import Std.Data.HashMap.Basic
 import Std.Data.RBMap.Basic
 import Mathlib.Lean.Meta.Basic
 import Mathlib.Mathport.Rename
+import Lean.Meta.Tactic.Rfl
```
Comment (semorrison) [inline Mathlib/Tactic/CC/Datatypes.lean:12]:
```suggestion
```

File: Mathlib/Data/Finset/Basic.lean (hunk 5 above)
Comment (semorrison) [reply Mathlib/Data/Finset/Basic.lean:2621]: https://github.com/leanprover-community/mathlib4/pull/12576

File: Mathlib/Data/Fintype/BigOperators.lean [hunk 13]
```diff
@@ -175,11 +175,11 @@ lemma card_filter_piFinset_const (s : Finset κ) (i : ι) (j : κ) :
...
 end Pi
 
--- FIXME ouch, this should be in the main file.
-@[simp]
+-- TODO: this is a basic thereom about `Fintype.card`,
+-- and ideally could be moved to `Mathlib.Data.Fintype.Card`.
 theorem Fintype.card_fun [DecidableEq α] [Fintype α] [Fintype β] :
     Fintype.card (α → β) = Fintype.card β ^ Fintype.card α := by
-  rw [Fintype.card_pi, Finset.prod_const]; rfl
+  simp
```
Comment (semorrison) [inline Mathlib/Data/Fintype/BigOperators.lean:182]: https://github.com/leanprover-community/mathlib4/pull/12577

File: lakefile.lean [hunk 14]
```diff
@@ -26,12 +26,12 @@ package mathlib where
...
-require std from git "https://github.com/leanprover/std4" @ "main"
+require std from git "https://github.com/leanprover/std4" @ "nightly-testing"
 require Qq from git "https://github.com/leanprover-community/quote4" @ "master"
-require aesop from git "https://github.com/leanprover-community/aesop" @ "master"
-require proofwidgets from git "https://github.com/leanprover-community/ProofWidgets4" @ "v0.0.30"
+require aesop from git "https://github.com/leanprover-community/aesop" @ "nightly-testing"
+require proofwidgets from git "https://github.com/leanprover-community/ProofWidgets4" @ "v0.0.35-pre"
 require Cli from git "https://github.com/leanprover/lean4-cli" @ "main"
-require importGraph from git "https://github.com/leanprover-community/import-graph.git" @ "main"
+require importGraph from git "https://github.com/leanprover-community/import-graph.git" @ "nightly-testing"
```
Comment (Ruben-VandeVelde) [inline lakefile.lean:34]: Don't forget to revert this

**Comment by semorrison** [comment]
!bench

**Comment by semorrison** [comment]
Benchmarking results are pretty discouraging.

File: Mathlib/Data/Rat/Lemmas.lean [hunk 15]
```diff
@@ -62,13 +62,16 @@ theorem num_den_mk {q : ℚ} {n d : ℤ} (hd : d ≠ 0) (qdf : q = n /. d) :
...
+  conv_rhs => rw [← Nat.cast_one, ← Nat.cast_add, Int.natAbs_cast]
 #align rat.num_mk Rat.num_mk
 
 theorem den_mk (n d : ℤ) : (n /. d).den = if d = 0 then 1 else d.natAbs / n.gcd d := by
   rcases d with ((_ | _) | _) <;>
-    simp [divInt, mkRat, Rat.normalize, Nat.succPNat, Int.sign, Int.gcd, -Nat.cast_succ]
+    simp [divInt, mkRat, Rat.normalize, Nat.succPNat, Int.sign, Int.gcd,
+      if_neg (Nat.cast_add_one_ne_zero _)]
+  conv_rhs => rw [← Nat.cast_one, ← Nat.cast_add, Int.natAbs_cast]
 #align rat.denom_mk Rat.den_mk
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Rat/Lemmas.lean:75]: We could avoid the non-terminal simp with
```lean
theorem num_mk (n d : ℤ) : (n /. d).num = d.sign * n / n.gcd d := by
  have (m : ℕ) : Int.natAbs (m + 1) = m + 1 := by
    rw [← Nat.cast_one, ← Nat.cast_add, Int.natAbs_cast]
  rcases d with ((_ | _) | _) <;>
  rw [← Int.div_eq_ediv_of_dvd] <;>
  simp [divInt, mkRat, Rat.normalize, Nat.succPNat, Int.sign, Int.gcd,
    Int.zero_ediv, Int.ofNat_dvd_left, Nat.gcd_dvd_left, this]
#align rat.num_mk Rat.num_mk

theorem den_mk (n d : ℤ) : (n /. d).den = if d = 0 then 1 else d.natAbs / n.gcd d := by
  have (m : ℕ) : Int.natAbs (m + 1) = m + 1 := by
    rw [← Nat.cast_one, ← Nat.cast_add, Int.natAbs_cast]
  rcases d with ((_ | _) | _) <;>
    simp [divInt, mkRat, Rat.normalize, Nat.succPNat, Int.sign, Int.gcd,
      if_neg (Nat.cast_add_one_ne_zero _), this]
#align rat.denom_mk Rat.den_mk
```

File: Mathlib/GroupTheory/CoprodI.lean [hunk 16]
```diff
@@ -581,7 +581,7 @@ theorem mem_smul_iff {i j : ι} {m₁ : M i} {m₂ : M j} {w : Word M} :
         · simp
         · simp (config := {contextual := true}) [Sigma.ext_iff]
   · rcases w with ⟨_ | _, _, _⟩ <;>
-    simp [or_comm, hij, Ne.symm hij, eq_comm]
+    simp [or_comm, hij, Ne.symm hij]; rw [eq_comm]
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/CoprodI.lean:584]: Could also be (a bit disappointingly):
```suggestion
    simp [or_comm, hij, Ne.symm hij, eq_comm (a := (⟨i, m₁⟩ : (i : ι) × M i))]
```

File: Mathlib/Data/Rat/Lemmas.lean (hunk 15 above)
Comment (semorrison) [reply Mathlib/Data/Rat/Lemmas.lean:75]: I've stuck this into #12595

**Comment by Parcly-Taxel** [comment]
> Benchmarking results are pretty discouraging.

Is it pre-existing defeq abuse that is causing the slowdown?

**Comment by semorrison** [comment]
> > Benchmarking results are pretty discouraging.
> 
> Is it pre-existing defeq abuse that is causing the slowdown?

Yes. It's just got worse in many cases.

File: Mathlib/GroupTheory/CoprodI.lean (hunk 16 above)
Comment (semorrison) [reply Mathlib/GroupTheory/CoprodI.lean:584]: Hmm, no thanks. :-)

**Comment by semorrison** [comment]
bors p=10

**Pull request closed without merging**